pycbc/waveform/waveform_modes.py

```diff
@@ -0,0 +1,491 @@
+# Copyright (C) 2020  Collin Capano, Alex Nitz
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation; either version 3 of the License, or (at your
+# option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+"""Provides functions and utilities for generating waveforms mode-by-mode.
+"""
+
+from string import Formatter
+import lal
+import lalsimulation
+
+from pycbc.types import (TimeSeries, FrequencySeries)
+from .waveform import (props, _check_lal_pars, check_args)
+from . import parameters
+
+
+def _formatdocstr(docstr):
+    """Utility for formatting docstrings with parameter information.
+    """
+    return docstr.format(
+        **{_p[1]: getattr(parameters, _p[1]).docstr(
+            prefix="    ", include_label=False).lstrip(' ')
+           for _p in Formatter().parse(docstr) if _p[1] is not None
+           })
+
+
+def _formatdocstrlist(docstr, paramlist, skip_params=None):
+    """Utility for formatting docstrings with parameter information.
+    """
+    if skip_params is None:
+        skip_params = []
+    pl = '\n'.join([_p.docstr(prefix="    ", include_label=False)
+                    for _p in paramlist if _p not in skip_params]).lstrip(' ')
+    return docstr.format(params=pl)
+
+
+def sum_modes(hlms, inclination, phi):
+    """Applies spherical harmonics and sums modes to produce a plus and cross
+    polarization.
+
+    Parameters
+    ----------
+    hlms : dict
+        Dictionary of ``(l, m)`` -> complex ``hlm``. The ``hlm`` may be a
+        complex number or array, or complex ``TimeSeries``. All modes in the
+        dictionary will be summed.
+    inclination : float
+        The inclination to use.
+    phi : float
+        The phase to use.
+
+    Returns
+    -------
+    complex float or array
+        The plus and cross polarization as a complex number. The real part
+        gives the plus, the negative imaginary part the cross.
+    """
+    out = None
+    for mode in hlms:
+        ell, m = mode
```

Comment:
Looks a bit strange to use `ell` for l and `m` for m. I think `l` and `m` is fine here, but if you want to appease CodeClimate `l_mode` and `m_mode` might be usable?

Reply:
It was habit from LaTeX. Switched to `l` (eat it, CodeClimate!).